File: gs_to_rhino.py
"""
3D Gaussian Splat PLY → Rhino import 가능한 컬러 PointCloud PLY 변환기.

입력: 3DGS 학습 결과 point_cloud.ply (x, y, z, f_dc_0..2, opacity, scale, rot 등)
출력: 표준 컬러 PLY (x, y, z, red, green, blue) — Rhino `_Import` 로 바로 읽힘

외부 의존성: numpy 만 필요 (plyfile 불필요, binary PLY를 직접 파싱).
"""

import numpy as np
import struct
import sys
import os

SH_C0 = 0.28209479177387814  # Spherical Harmonics 0차 계수


def parse_header(f):
    """Text header를 읽어서 (vertex_count, property_names, ascii_or_binary) 반환."""
    line = f.readline().strip()
    if line != b'ply':
        raise ValueError(f'PLY 매직이 아님: {line!r}')

    vertex_count = None
    props = []  # (dtype_str, name)
    fmt = None

    while True:
        line = f.readline()
        if not line:
            raise ValueError('unexpected EOF in header')
        s = line.strip()
        if s == b'end_header':
            break
        parts = s.split()
        if len(parts) == 0:
            continue
        key = parts[0]
        if key == b'format':
            fmt = parts[1].decode('ascii')
        elif key == b'element' and parts[1] == b'vertex':
            vertex_count = int(parts[2])
        elif key == b'property':
            # property <type> <name>
            dtype_str = parts[1].decode('ascii')
            name = parts[2].decode('ascii')
            props.append((dtype_str, name))
        # comment / element face 등은 무시

    if vertex_count is None:
        raise ValueError('vertex element를 찾지 못함')
    if fmt != 'binary_little_endian':
        raise ValueError(
            f'이 스크립트는 binary_little_endian 만 지원. 현재: {fmt}'
        )

    return vertex_count, props


def ply_type_to_numpy(t):
    return {
        'float': '<f4', 'float32': '<f4',
        'double': '<f8', 'float64': '<f8',
        'uchar': '<u1', 'uint8': '<u1',
        'char': '<i1', 'int8': '<i1',
        'ushort': '<u2', 'uint16': '<u2',
        'short': '<i2', 'int16': '<i2',
        'uint': '<u4', 'uint32': '<u4',
        'int': '<i4', 'int32': '<i4',
    }[t]


def apply_viewer_model_matrix(xyz):
    """뷰어 viewer.html 의 modelMatrix 와 동일한 좌표 변환.

    modelMatrix = [[-1, 0, 0, 0],
                   [ 0, 0, 1, 0],
                   [ 0, 1, 0, 0],
                   [ 0, 0, 0, 1]]

    즉 (x, y, z) -> (-x, z, y). X 반전 + Y <-> Z 스왑.
    Rhino 에서 import 한 결과가 viewer 와 완전히 같은 방향/위치로 보이도록 정렬.
    """
    out = np.empty_like(xyz)
    out[:, 0] = -xyz[:, 0]
    out[:, 1] = xyz[:, 2]
    out[:, 2] = xyz[:, 1]
    return out


def convert(in_path, out_path, opacity_thr=0.1, max_points=None, seed=0,
            apply_viewer_matrix=True):
    size = os.path.getsize(in_path)
    print(f'[1/5] 입력: {in_path} ({size/1024/1024:.1f} MB)')

    with open(in_path, 'rb') as f:
        vcount, props = parse_header(f)
        header_end = f.tell()
        print(f'      vertex_count = {vcount:,}')
        print(f'      properties   = {len(props)} '
              f'({", ".join(p[1] for p in props[:6])}...)')

        # 구조화된 dtype 생성
        dtype = np.dtype([(name, ply_type_to_numpy(t)) for (t, name) in props])
        record_size = dtype.itemsize
        expected = vcount * record_size
        remaining = size - header_end
        if remaining < expected:
            raise ValueError(
                f'파일이 예상보다 작음: {remaining} < {expected} '
                f'(레코드 {record_size} bytes * {vcount})'
            )

        print(f'[2/5] 바이너리 페이로드 읽는 중 '
              f'({expected/1024/1024:.1f} MB)...')
        arr = np.frombuffer(f.read(expected), dtype=dtype, count=vcount)

    # 필수 필드 확인
    needed = ['x', 'y', 'z', 'f_dc_0', 'f_dc_1', 'f_dc_2', 'opacity']
    missing = [n for n in needed if n not in arr.dtype.names]
    if missing:
        raise ValueError(f'필수 필드 누락: {missing}')

    print('[3/5] SH → RGB 변환 + opacity 필터링...')
    # SH DC term → 선형 RGB → [0,1] clip → 8bit
    rgb = np.stack([arr['f_dc_0'], arr['f_dc_1'], arr['f_dc_2']], axis=-1)
    rgb = np.clip(0.5 + SH_C0 * rgb, 0.0, 1.0)
    rgb = (rgb * 255.0 + 0.5).astype(np.uint8)

    # opacity 로짓 → sigmoid (안정적 계산)
    op_raw = arr['opacity']
    op = np.where(
        op_raw >= 0,
        1.0 / (1.0 + np.exp(-op_raw)),
        np.exp(op_raw) / (1.0 + np.exp(op_raw)),
    )
    mask = op > opacity_thr

    xyz = np.stack([arr['x'], arr['y'], arr['z']], axis=-1)
    xyz = xyz[mask]
    rgb = rgb[mask]
    kept = len(xyz)
    print(f'      opacity > {opacity_thr} : {kept:,} / {vcount:,} '
          f'({kept/vcount*100:.1f}%)')

    if apply_viewer_matrix:
        print('      viewer modelMatrix 적용: (x, y, z) -> (-x, z, y)')
        xyz = apply_viewer_model_matrix(xyz)

    if max_points and kept > max_points:
        rng = np.random.default_rng(seed)
        idx = rng.choice(kept, max_points, replace=False)
        xyz = xyz[idx]
        rgb = rgb[idx]
        print(f'      서브샘플 → {len(xyz):,}')

    n = len(xyz)

    print(f'[4/5] 출력 작성 중: {out_path}')
    with open(out_path, 'wb') as f:
        header = (
            'ply\n'
            'format binary_little_endian 1.0\n'
            'comment converted from 3D Gaussian Splat PLY\n'
            'comment SH DC -> RGB, opacity filtered, ready for Rhino _Import\n'
            f'element vertex {n}\n'
            'property float x\n'
            'property float y\n'
            'property float z\n'
            'property uchar red\n'
            'property uchar green\n'
            'property uchar blue\n'
            'end_header\n'
        )
        f.write(header.encode('ascii'))

        # 레코드 단위로 interleave: (float x 3) + (uchar x 3) = 15 bytes
        out_dtype = np.dtype([
            ('x', '<f4'), ('y', '<f4'), ('z', '<f4'),
            ('r', '<u1'), ('g', '<u1'), ('b', '<u1'),
        ])
        rec = np.empty(n, dtype=out_dtype)
        rec['x'] = xyz[:, 0]
        rec['y'] = xyz[:, 1]
        rec['z'] = xyz[:, 2]
        rec['r'] = rgb[:, 0]
        rec['g'] = rgb[:, 1]
        rec['b'] = rgb[:, 2]
        f.write(rec.tobytes())

    out_size = os.path.getsize(out_path)
    print(f'[5/5] 완료! {n:,} points, {out_size/1024/1024:.1f} MB')
    bbox_min = xyz.min(axis=0)
    bbox_max = xyz.max(axis=0)
    print(f'      bbox min = {bbox_min}')
    print(f'      bbox max = {bbox_max}')
    print(f'      size (W×D×H) = {bbox_max - bbox_min}')


if __name__ == '__main__':
    # 기본 입출력 경로 (인자 없을 때)
    default_in = os.path.join(os.path.dirname(__file__), 'point_cloud.ply')
    default_out = os.path.join(os.path.dirname(__file__), 'point_cloud_rhino.ply')

    in_path = sys.argv[1] if len(sys.argv) > 1 else default_in
    out_path = sys.argv[2] if len(sys.argv) > 2 else default_out

    convert(
        in_path,
        out_path,
        opacity_thr=0.1,          # 낮은 opacity 제거 (플로터 제거)
        max_points=800_000,       # Rhino 쾌적하게 다루는 상한
        apply_viewer_matrix=True, # viewer.html 의 modelMatrix 와 같은 방향으로 정렬
    )

File: modules/utils/gs-to-rhino.js
// modules/utils/gs-to-rhino.js
// -----------------------------------------------------------------------------
// 3D Gaussian Splat PLY → Rhino import 가능한 컬러 PointCloud PLY 변환 (JS 포팅).
// gs_to_rhino.py 와 동일 로직을 브라우저 환경에서 수행.
//
// 입력: 3DGS 학습 결과 point_cloud.ply (x, y, z, f_dc_0..2, opacity, ...)
// 출력: 표준 컬러 PLY (x, y, z, red, green, blue) 의 ArrayBuffer
//        → Blob 으로 감싸 다운로드.
//
// 변환 단계:
//   1) PLY 헤더 파싱 (text)
//   2) binary_little_endian 페이로드 → interleaved record → 필드별 추출
//   3) SH DC → sRGB (0.5 + SH_C0 * dc, clip [0,1], *255)
//   4) sigmoid(opacity) > 0.1 필터
//   5) (x, y, z) → (-x, z, y) viewer modelMatrix 적용
//   6) 최대 max_points 로 서브샘플 (기본 800,000)
//   7) 표준 컬러 PLY 직렬화
// -----------------------------------------------------------------------------

const SH_C0 = 0.28209479177387814;

// PLY property type → (byteSize, readerName) 매핑. DataView 기준.
const TYPE_INFO = {
    'float':   { size: 4, reader: 'getFloat32' },
    'float32': { size: 4, reader: 'getFloat32' },
    'double':  { size: 8, reader: 'getFloat64' },
    'float64': { size: 8, reader: 'getFloat64' },
    'uchar':   { size: 1, reader: 'getUint8'  },
    'uint8':   { size: 1, reader: 'getUint8'  },
    'char':    { size: 1, reader: 'getInt8'   },
    'int8':    { size: 1, reader: 'getInt8'   },
    'ushort':  { size: 2, reader: 'getUint16' },
    'uint16':  { size: 2, reader: 'getUint16' },
    'short':   { size: 2, reader: 'getInt16'  },
    'int16':   { size: 2, reader: 'getInt16'  },
    'uint':    { size: 4, reader: 'getUint32' },
    'uint32':  { size: 4, reader: 'getUint32' },
    'int':     { size: 4, reader: 'getInt32'  },
    'int32':   { size: 4, reader: 'getInt32'  },
};

/**
 * ArrayBuffer → { vertexCount, props: [{name,type,offset,size,reader}], payloadOffset }
 * 헤더는 ASCII 라인 기반. binary_little_endian 만 지원.
 */
function parseHeader(buffer) {
    const bytes = new Uint8Array(buffer);
    // 헤더 끝 위치 찾기 ("end_header\n")
    let headerEnd = -1;
    const MAX_HEADER = Math.min(bytes.length, 64 * 1024); // 헤더는 수 KB 이내
    const endMarker = new TextEncoder().encode('end_header\n');
    outer: for (let i = 0; i < MAX_HEADER - endMarker.length; i++) {
        for (let j = 0; j < endMarker.length; j++) {
            if (bytes[i + j] !== endMarker[j]) continue outer;
        }
        headerEnd = i + endMarker.length;
        break;
    }
    if (headerEnd < 0) throw new Error('PLY end_header 를 찾지 못함 — 올바른 PLY 파일이 아닌 것 같습니다.');

    const headerText = new TextDecoder('ascii').decode(bytes.subarray(0, headerEnd));
    const lines = headerText.split(/\r?\n/);
    if (lines[0].trim() !== 'ply') throw new Error('PLY magic 아님: ' + lines[0]);

    let fmt = null, vertexCount = null;
    const props = []; // {type, name, offset, size, reader}
    let curOffset = 0;
    for (const rawLine of lines) {
        const s = rawLine.trim();
        if (!s || s === 'end_header') continue;
        const parts = s.split(/\s+/);
        const key = parts[0];
        if (key === 'format') {
            fmt = parts[1];
        } else if (key === 'element' && parts[1] === 'vertex') {
            vertexCount = parseInt(parts[2], 10);
        } else if (key === 'property') {
            const type = parts[1];
            const name = parts[2];
            const info = TYPE_INFO[type];
            if (!info) throw new Error(`지원 안 되는 PLY property type: ${type}`);
            props.push({ type, name, offset: curOffset, size: info.size, reader: info.reader });
            curOffset += info.size;
        }
    }
    if (fmt !== 'binary_little_endian 1.0' && fmt !== 'binary_little_endian') {
        throw new Error(`binary_little_endian 만 지원. 현재: ${fmt}`);
    }
    if (vertexCount === null) throw new Error('vertex element 를 찾지 못함');
    return { vertexCount, props, payloadOffset: headerEnd, recordSize: curOffset };
}

/** 진행률 로깅을 위한 소단위 마이크로태스크 양보 */
const yieldNow = () => new Promise(r => setTimeout(r, 0));

/**
 * GS PLY ArrayBuffer 를 Rhino 컬러 PLY ArrayBuffer 로 변환.
 *
 * @param {ArrayBuffer} inputBuffer
 * @param {object} [opts]
 *   opts.opacityThreshold   (기본 0.1)  — sigmoid(opacity) 이 값보다 크면 keep
 *   opts.maxPoints          (기본 800_000) — 서브샘플 상한. null 이면 제한 없음
 *   opts.applyViewerMatrix  (기본 true) — (x,y,z) → (-x,z,y) 적용
 *   opts.onProgress         (기본 noop) — ({stage, pct, msg}) 콜백
 * @returns {Promise<ArrayBuffer>}  Rhino PLY 바이너리
 */
export async function convertGsPlyToRhinoPly(inputBuffer, opts = {}) {
    const {
        opacityThreshold = 0.1,
        maxPoints = 800_000,
        applyViewerMatrix = true,
        onProgress = () => {},
    } = opts;

    onProgress({ stage: 'parse', pct: 0, msg: 'PLY 헤더 파싱 중…' });
    const { vertexCount, props, payloadOffset, recordSize } = parseHeader(inputBuffer);

    // 필수 필드 탐색
    const need = ['x', 'y', 'z', 'f_dc_0', 'f_dc_1', 'f_dc_2', 'opacity'];
    const byName = {};
    for (const p of props) byName[p.name] = p;
    const missing = need.filter(n => !byName[n]);
    if (missing.length) throw new Error('필수 필드 누락: ' + missing.join(', '));

    const payload = new DataView(inputBuffer, payloadOffset, vertexCount * recordSize);
    const totalBytes = vertexCount * recordSize;
    onProgress({ stage: 'read', pct: 5, msg: `vertex = ${vertexCount.toLocaleString()} / 레코드 ${recordSize}B` });

    // 각 필드 reader 함수
    const readField = (i, fieldName) => {
        const p = byName[fieldName];
        return payload[p.reader](i * recordSize + p.offset, true /* little endian */);
    };

    // 1) 필터 + 변환 — 스트리밍으로 진행 (vertex 가 많으면 수천만 개이므로 chunk 처리)
    const xyzOut = new Float32Array(vertexCount * 3); // 최대 크기로 할당, 끝에 자름
    const rgbOut = new Uint8Array(vertexCount * 3);
    let kept = 0;

    const CHUNK = 200_000;
    for (let chunkStart = 0; chunkStart < vertexCount; chunkStart += CHUNK) {
        const end = Math.min(chunkStart + CHUNK, vertexCount);
        for (let i = chunkStart; i < end; i++) {
            // opacity sigmoid (수치 안정 형태)
            const op = readField(i, 'opacity');
            let sig;
            if (op >= 0) {
                const e = Math.exp(-op);
                sig = 1 / (1 + e);
            } else {
                const e = Math.exp(op);
                sig = e / (1 + e);
            }
            if (sig <= opacityThreshold) continue;

            // SH DC → RGB
            const d0 = readField(i, 'f_dc_0');
            const d1 = readField(i, 'f_dc_1');
            const d2 = readField(i, 'f_dc_2');
            let r = 0.5 + SH_C0 * d0;
            let g = 0.5 + SH_C0 * d1;
            let b = 0.5 + SH_C0 * d2;
            if (r < 0) r = 0; else if (r > 1) r = 1;
            if (g < 0) g = 0; else if (g > 1) g = 1;
            if (b < 0) b = 0; else if (b > 1) b = 1;

            // 위치 + viewer modelMatrix
            let x = readField(i, 'x');
            let y = readField(i, 'y');
            let z = readField(i, 'z');
            if (applyViewerMatrix) {
                const nx = -x, ny = z, nz = y;
                x = nx; y = ny; z = nz;
            }

            const k = kept * 3;
            xyzOut[k]     = x;
            xyzOut[k + 1] = y;
            xyzOut[k + 2] = z;
            rgbOut[k]     = (r * 255 + 0.5) | 0;
            rgbOut[k + 1] = (g * 255 + 0.5) | 0;
            rgbOut[k + 2] = (b * 255 + 0.5) | 0;
            kept++;
        }
        onProgress({
            stage: 'filter',
            pct: 5 + Math.round(80 * end / vertexCount),
            msg: `필터 진행: ${end.toLocaleString()} / ${vertexCount.toLocaleString()} (keep ${kept.toLocaleString()})`,
        });
        await yieldNow(); // UI 블록 방지
    }

    // 2) 서브샘플 (선택)
    let nFinal = kept;
    let idxMap = null;
    if (maxPoints && kept > maxPoints) {
        // Fisher-Yates 로 상위 maxPoints 개 인덱스 선택
        idxMap = new Uint32Array(kept);
        for (let i = 0; i < kept; i++) idxMap[i] = i;
        for (let i = kept - 1; i > 0; i--) {
            const j = (Math.random() * (i + 1)) | 0;
            const t = idxMap[i]; idxMap[i] = idxMap[j]; idxMap[j] = t;
        }
        nFinal = maxPoints;
        onProgress({ stage: 'sample', pct: 88, msg: `서브샘플 → ${nFinal.toLocaleString()}` });
    }

    // 3) 출력 PLY 빌드
    onProgress({ stage: 'write', pct: 92, msg: '출력 바이너리 작성…' });
    const headerText =
        'ply\n' +
        'format binary_little_endian 1.0\n' +
        'comment converted from 3D Gaussian Splat PLY\n' +
        'comment SH DC -> RGB, opacity filtered, ready for Rhino _Import\n' +
        `element vertex ${nFinal}\n` +
        'property float x\n' +
        'property float y\n' +
        'property float z\n' +
        'property uchar red\n' +
        'property uchar green\n' +
        'property uchar blue\n' +
        'end_header\n';
    const headerBytes = new TextEncoder().encode(headerText);

    const recSize = 15; // 4*3 + 1*3
    const out = new ArrayBuffer(headerBytes.byteLength + nFinal * recSize);
    new Uint8Array(out, 0, headerBytes.byteLength).set(headerBytes);
    const dv = new DataView(out, headerBytes.byteLength);
    for (let i = 0; i < nFinal; i++) {
        const src = idxMap ? idxMap[i] : i;
        const s3 = src * 3;
        const o = i * recSize;
        dv.setFloat32(o,     xyzOut[s3],     true);
        dv.setFloat32(o + 4, xyzOut[s3 + 1], true);
        dv.setFloat32(o + 8, xyzOut[s3 + 2], true);
        dv.setUint8(o + 12, rgbOut[s3]);
        dv.setUint8(o + 13, rgbOut[s3 + 1]);
        dv.setUint8(o + 14, rgbOut[s3 + 2]);
    }
    onProgress({ stage: 'done', pct: 100, msg: `완료: ${nFinal.toLocaleString()} points` });
    return out;
}
